File: src/instance.ts
/**
 * createShield(config) — an instance with its own configuration, sinks and
 * event buffer.
 *
 * Everything in shield.ts is module-level: one bus, one salt, one announced
 * set. That is fine for a single app on a laptop and wrong for a library:
 * two tenants in one process, a test that needs a clean slate, an app that
 * ships events to a webhook instead of a home-directory file. A Shield
 * instance owns:
 *
 *   - its app label, thresholds, canary, output/tool policy, registered secrets
 *   - its own subscribers and recent-event buffer
 *   - the decision whether to forward events to the global bus (default yes,
 *     so `shield logs` / `shield status` keep working) and whether to attach
 *     the JSONL file sink under Node (default yes)
 *
 * The module-level functions keep working unchanged; `getDefaultShield()`
 * is the instance they are equivalent to.
 */

import {
  SHIELD_VERSION,
  announceShield,
  detectInjection,
  emitShieldEvent,
  hardenSystemPrompt,
  outputLeakedCanary,
  scanDetail,
  wrapUntrusted,
  type HardenResult,
  type InjectionScan,
  type ShieldEvent,
  type SubscribeOptions,
} from "./shield.js";
import {
  evaluateToolCall,
  outputDetail,
  scanOutput,
  type OutputCategory,
  type OutputScan,
  type OutputScanContext,
  type ToolCall,
  type ToolCallContext,
  type ToolDecision,
  type ToolPolicy,
} from "./output.js";
import { initFileLogger } from "./logger.js";

export type Sink = (event: ShieldEvent) => void;

export interface OutputConfig {
  /** Hosts the model may legitimately reference in URLs. */
  allowedHosts?: string[];
  /** Per-category on/off (all on by default). */
  detectors?: Partial<Record<OutputCategory, boolean>>;
  /** Flag threshold for output scans (default 0.5). */
  threshold?: number;
}

export interface ShieldConfig {
  /** App label used as the event source. */
  app?: string;
  /** Input-scan flag threshold (default 0.5). */
  threshold?: number;
  /** Pin the canary token (see README: prompt caching across workers). */
  canary?: string;
  /** What to keep in `detail`: the triage excerpt (default), a 32-bit hash of
   *  it, or nothing. Excerpts contain snippets of user content. */
  redact?: "excerpt" | "hash" | "none";
  /** Values the model must never emit. Never logged. */
  secrets?: string[];
  output?: OutputConfig;
  toolPolicy?: ToolPolicy;
  /** Extra event sinks (console, webhook, OpenTelemetry…). */
  sinks?: Sink[];
  /** Also emit on the module-level bus (default true) so `shield logs` and
   *  existing onShieldEvent subscribers see this instance's events. */
  forwardToGlobal?: boolean;
  /** Attach the JSONL file sink under Node (default true). */
  fileLogger?: boolean;
  /** Print the startup banner (default true; SHIELD_QUIET=1 also silences). */
  banner?: boolean;
}

export interface ScanInputOptions {
  /** Channel qualifier appended to the source: "tool_result", "document"… */
  channel?: string;
  /** Override the instance threshold for this call. */
  threshold?: number;
}

function hash32hex(s: string): string {
  let h = 0x811c9dc5;
  for (let i = 0; i < s.length; i++) { h ^= s.charCodeAt(i); h = Math.imul(h, 16777619) >>> 0; }
  return h.toString(16).padStart(8, "0");
}

export class Shield {
  readonly config: Readonly<ShieldConfig>;
  readonly app: string;
  private readonly handlers: Sink[] = [];
  private readonly recent: ShieldEvent[] = [];
  private static readonly RECENT_MAX = 256;

  constructor(config: ShieldConfig = {}) {
    this.config = Object.freeze({ ...config });
    this.app = config.app ?? "shield";
    for (const s of config.sinks ?? []) this.handlers.push(s);
    if (config.fileLogger !== false) void initFileLogger();
    announceShield({ appLabel: this.app, banner: config.banner });
  }

  get version(): string {
    return SHIELD_VERSION;
  }

  // ── events ────────────────────────────────────────────────────────────────

  /** Subscribe to this instance's events. Returns an unsubscribe function. */
  on(handler: Sink, opts: SubscribeOptions = {}): () => void {
    if (opts.replay) for (const ev of [...this.recent]) { try { handler(ev); } catch (err) { console.error("[shield] sink threw during replay; continuing:", err); } }
    this.handlers.push(handler);
    return () => { const i = this.handlers.indexOf(handler); if (i >= 0) this.handlers.splice(i, 1); };
  }

  /** Events this instance has emitted (most recent 256). */
  get events(): readonly ShieldEvent[] {
    return this.recent;
  }

  emit(event: Omit<ShieldEvent, "timestamp">): ShieldEvent {
    const full: ShieldEvent = { ...event, detail: this.redact(event.detail), timestamp: new Date().toISOString() };
    this.recent.push(full);
    if (this.recent.length > Shield.RECENT_MAX) this.recent.splice(0, this.recent.length - Shield.RECENT_MAX);
    for (const h of [...this.handlers]) {
      try { h(full); } catch (err) { console.error("[shield] sink threw; continuing:", err); }
    }
    if (this.config.forwardToGlobal !== false) {
      const { timestamp: _t, ...rest } = full;
      void _t;
      emitShieldEvent(rest);
    }
    return full;
  }

  private redact(detail: string): string {
    switch (this.config.redact) {
      case "hash": return detail ? `sha:${hash32hex(detail)}` : "";
      case "none": return "";
      default: return detail;
    }
  }

  private source(channel?: string): string {
    return channel ? `${this.app}:${channel}` : this.app;
  }

  // ── input side ────────────────────────────────────────────────────────────

  /** Scan untrusted text; emits `injection_detected` when flagged. */
  scanInput(text: string, opts: ScanInputOptions = {}): InjectionScan {
    const scan = detectInjection(text, opts.threshold ?? this.config.threshold ?? 0.5);
    if (scan.flagged) {
      this.emit({
        type: "injection_detected",
        source: this.source(opts.channel),
        detail: scanDetail(text, scan),
        score: scan.score,
        patterns: scan.matches,
        direction: "input",
      });
    }
    return scan;
  }

  wrap(content: string, label: string): string {
    return wrapUntrusted(content, label);
  }

  harden(base: string): HardenResult {
    return hardenSystemPrompt(base, this.config.canary);
  }

  // ── output side ───────────────────────────────────────────────────────────

  /**
   * Scan model output for leaked secrets, PII, exfiltration channels, echoed
   * injections and the canary. Emits `canary_leaked` for a canary hit (kept
   * separate so `shield status` counts it as before) and `output_flagged`
   * for everything else.
   */
  scanOutput(text: string, ctx: OutputScanContext = {}): OutputScan {
    const merged: OutputScanContext = {
      allowedHosts: this.config.output?.allowedHosts,
      detectors: this.config.output?.detectors,
      threshold: this.config.output?.threshold,
      secrets: this.config.secrets,
      ...ctx,
    };
    const scan = scanOutput(text, merged);
    if (merged.canary && outputLeakedCanary(text, merged.canary)) {
      this.emit({ type: "canary_leaked", source: this.app, detail: text.slice(0, 200), direction: "output" });
      console.warn(`[shield] Canary leak detected in response from ${this.app}`);
    }
    const rest = scan.findings.filter((f) => f.category !== "canary");
    if (rest.length > 0 && scan.flagged) {
      this.emit({
        type: "output_flagged",
        source: this.app,
        detail: outputDetail({ ...scan, findings: rest }),
        score: scan.score,
        patterns: rest.map((f) => f.label),
        direction: "output",
      });
    }
    return scan;
  }

  /** Evaluate a requested tool call against the policy; emits `tool_call_gated` unless allowed. */
  checkToolCall(call: ToolCall, ctx: ToolCallContext = {}): ToolDecision {
    const d = evaluateToolCall(call, this.config.toolPolicy ?? {}, ctx);
    if (d.decision !== "allow") {
      this.emit({
        type: "tool_call_gated",
        source: this.app,
        detail: `${d.decision}: ${call.name} — ${d.reasons.join("; ")}`.slice(0, 300),
        score: d.decision === "block" ? 1 : 0.6,
        patterns: [`tool:${call.name}`, `decision:${d.decision}`],
        direction: "tool",
      });
    }
    return d;
  }
}

export function createShield(config: ShieldConfig = {}): Shield {
  return new Shield(config);
}

let defaultInstance: Shield | null = null;

/** The instance the module-level API is equivalent to (app "shield", defaults). */
export function getDefaultShield(): Shield {
  if (!defaultInstance) defaultInstance = new Shield({ banner: false });
  return defaultInstance;
}

File: python/shield/instance.py
"""
create_shield(...) — an instance with its own configuration, sinks and event
buffer. Mirrors src/instance.ts.

The module-level functions keep working unchanged; `get_default_shield()` is
the instance they are equivalent to. Events are forwarded to the module bus
(and therefore the shared JSONL log) unless `forward_to_global=False`.
"""
from __future__ import annotations

from collections import deque
from datetime import datetime, timezone
from typing import Any, Callable, Optional, Sequence

from .core import SHIELD_VERSION, announce_shield, harden_system_prompt, output_leaked_canary, wrap_untrusted
from .detect import InjectionScan, detect_injection, scan_detail
from .logger import emit_event
from .output import (
    OutputScan,
    ToolCall,
    ToolDecision,
    ToolPolicy,
    evaluate_tool_call,
    output_detail,
    scan_output,
)

Sink = Callable[[dict], None]


def _hash32hex(s: str) -> str:
    h = 0x811C9DC5
    b = s.encode("utf-16-le")
    for i in range(0, len(b), 2):
        h ^= b[i] | (b[i + 1] << 8)
        h = (h * 16777619) & 0xFFFFFFFF
    return f"{h:08x}"


class Shield:
    """See create_shield()."""

    RECENT_MAX = 256

    def __init__(
        self,
        *,
        app: str = "shield",
        threshold: float = 0.5,
        canary: Optional[str] = None,
        redact: str = "excerpt",               # "excerpt" | "hash" | "none"
        secrets: Optional[Sequence[str]] = None,
        allowed_hosts: Optional[Sequence[str]] = None,
        output_detectors: Optional[dict[str, bool]] = None,
        output_threshold: float = 0.5,
        tool_policy: Optional[ToolPolicy] = None,
        sinks: Optional[Sequence[Sink]] = None,
        forward_to_global: bool = True,
        banner: bool = True,
    ):
        self.app = app
        self.threshold = threshold
        self.canary = canary
        self.redact = redact
        self.secrets = list(secrets or [])
        self.allowed_hosts = list(allowed_hosts) if allowed_hosts is not None else None
        self.output_detectors = dict(output_detectors or {})
        self.output_threshold = output_threshold
        self.tool_policy = tool_policy or ToolPolicy()
        self.forward_to_global = forward_to_global
        self._handlers: list[Sink] = list(sinks or [])
        self._recent: deque = deque(maxlen=self.RECENT_MAX)
        announce_shield(app, banner=banner)

    @property
    def version(self) -> str:
        return SHIELD_VERSION

    # ── events ────────────────────────────────────────────────────────────────

    def on(self, handler: Sink, *, replay: bool = False) -> Callable[[], None]:
        if replay:
            for ev in list(self._recent):
                try:
                    handler(ev)
                except Exception as err:  # noqa: BLE001
                    print(f"[shield] sink threw during replay; continuing: {err!r}")
        self._handlers.append(handler)

        def off() -> None:
            try:
                self._handlers.remove(handler)
            except ValueError:
                pass

        return off

    @property
    def events(self) -> list[dict]:
        return list(self._recent)

    def _redact(self, detail: str) -> str:
        if self.redact == "hash":
            return f"sha:{_hash32hex(detail)}" if detail else ""
        if self.redact == "none":
            return ""
        return detail

    def emit(
        self,
        type: str,
        *,
        source: Optional[str] = None,
        detail: str = "",
        score: Optional[float] = None,
        patterns: Optional[list[str]] = None,
        direction: Optional[str] = None,
    ) -> dict:
        detail = self._redact(detail)
        event: dict = {
            "type": type,
            "source": source or self.app,
            "detail": detail[:300],
            "timestamp": datetime.now(timezone.utc).isoformat(),
        }
        if score is not None:
            event["score"] = round(score, 4)
        if patterns is not None:
            event["patterns"] = patterns
        if direction:
            event["direction"] = direction
        self._recent.append(event)
        for h in list(self._handlers):
            try:
                h(event)
            except Exception as err:  # noqa: BLE001
                print(f"[shield] sink threw; continuing: {err!r}")
        if self.forward_to_global:
            emit_event(type, source=event["source"], detail=detail, score=score,  # type: ignore[arg-type]
                       patterns=patterns, direction=direction)
        return event

    def _source(self, channel: Optional[str]) -> str:
        return f"{self.app}:{channel}" if channel else self.app

    # ── input side ────────────────────────────────────────────────────────────

    def scan_input(self, text: str, *, channel: Optional[str] = None, threshold: Optional[float] = None) -> InjectionScan:
        scan = detect_injection(text, threshold if threshold is not None else self.threshold)
        if scan.flagged:
            self.emit("injection_detected", source=self._source(channel), detail=scan_detail(text, scan),
                      score=scan.score, patterns=scan.matches, direction="input")
        return scan

    def wrap(self, content: str, label: str) -> str:
        return wrap_untrusted(content, label)

    def harden(self, base: str) -> tuple[str, str]:
        return harden_system_prompt(base, self.canary)

    # ── output side ───────────────────────────────────────────────────────────

    def scan_output(
        self,
        text: str,
        *,
        canary: Optional[str] = None,
        input_scans: Optional[Sequence[InjectionScan]] = None,
        allowed_hosts: Optional[Sequence[str]] = None,
        detectors: Optional[dict[str, bool]] = None,
        threshold: Optional[float] = None,
    ) -> OutputScan:
        scan = scan_output(
            text,
            canary=canary,
            input_scans=input_scans,
            allowed_hosts=allowed_hosts if allowed_hosts is not None else self.allowed_hosts,
            detectors=detectors if detectors is not None else self.output_detectors,
            threshold=threshold if threshold is not None else self.output_threshold,
            secrets=self.secrets,
        )
        if canary and output_leaked_canary(text, canary):
            self.emit("canary_leaked", detail=text[:200], direction="output")
            print(f"[shield] WARNING: canary leaked in response from {self.app}")
        rest = [f for f in scan.findings if f.category != "canary"]
        if rest and scan.flagged:
            self.emit("output_flagged", detail=output_detail(OutputScan(scan.score, scan.flagged, rest)),
                      score=scan.score, patterns=[f.label for f in rest], direction="output")
        return scan

    def check_tool_call(
        self,
        call: ToolCall,
        *,
        untrusted_input_seen: bool = False,
        input_scans: Optional[Sequence[InjectionScan]] = None,
    ) -> ToolDecision:
        d = evaluate_tool_call(call, self.tool_policy, untrusted_input_seen=untrusted_input_seen, input_scans=input_scans)
        if d.decision != "allow":
            self.emit("tool_call_gated",
                      detail=f"{d.decision}: {call.name} — {'; '.join(d.reasons)}",
                      score=1.0 if d.decision == "block" else 0.6,
                      patterns=[f"tool:{call.name}", f"decision:{d.decision}"], direction="tool")
        return d


def create_shield(**config: Any) -> Shield:
    return Shield(**config)


_default: Optional[Shield] = None


def get_default_shield() -> Shield:
    global _default
    if _default is None:
        _default = Shield(banner=False)
    return _default
